server/src/main/java/org/opensearch/index/mapper/SourceFieldMapper.java

```diff
@@ -119,21 +120,75 @@ public static class Builder extends MetadataFieldMapper.Builder {
             Collections.emptyList()
         );
 
+        /**
+         * A mapping parameter which define whether the recovery_source should be added or not. Default value is true.
+         * <p>
+         * Recovery source gets added if source is disabled or there are filters that are applied on _source using
+         * {@link #includes}/{@link #excludes}, which has the possibility to change the original document provided by
+         * customer. Recovery source is not a permanent field and gets removed during merges. Refer this merge
+         * policy: org.opensearch.index.engine.RecoverySourcePruneMergePolicy
+         * <p>
+         * The main reason for adding the _recovery_source was to ensure Peer to Peer recovery if segments
+         * are not flushed to the disk. If you are disabling the recovery source, then ensure that you are calling
+         * flush operation of Opensearch periodically to ensure that segments are flushed to the disk and if required
+         * Peer to Peer recovery can happen using segment files rather than replaying traffic by querying Lucene
+         * snapshot.
+         *
+         * <p>
+         * This is an expert mapping parameter.
+         *
+         */
+        private final Parameter<Boolean> recoverySourceEnabled = Parameter.boolParam(
+            "recovery_source_enabled",
+            false,
+            m -> toType(m).recoverySourceEnabled,
+            Defaults.ENABLED
+        );
+
+        /**
+         * Provides capability to add specific fields in the recovery_source.
+         * <p>
+         * Refer {@link #recoverySourceEnabled} for more details
+         * This is an expert parameter.
+         */
+        private final Parameter<List<String>> recoverySourceIncludes = Parameter.stringArrayParam(
+            "recovery_source_includes",
+            false,
+            m -> Arrays.asList(toType(m).recoverySourceIncludes),
+            Collections.emptyList()
+        );
+
+        /**
+         * Provides capability to remove specific fields in the recovery_source.
+         *
+         * Refer {@link #recoverySourceEnabled} for more details
+         * This is an expert parameter.
+         */
+        private final Parameter<List<String>> recoverySourceExcludes = Parameter.stringArrayParam(
+            "recovery_source_excludes",
+            false,
+            m -> Arrays.asList(toType(m).recoverySourceExcludes),
+            Collections.emptyList()
+        );
```

Comment:
Do we foresee people benefitting from these extra settings?

I kind of feel like you either need recovery source (i.e. like most people, you're relying on automatic flush behavior and don't keep track of what's been flushed) or you don't (because you're sure that everything you've sent has been flushed). 

I'm not sure if I see the value of partial recovery source. Maybe I'm just not thinking about it properly, though.

Reply:
The reason I keep this is because I feel, people might have usecase where they want to remove 1 field which takes a lot of storage like vectors. So I can remove vectors from recovery source, source and then ensure all fields are present for recovery will ensure that I am not losing my data and I am just lossing vectors for few documents which I can easily ingest later on.

So it provides me benefit of saying what fields I don't want to loose and which fields I am okay to loose.

Reply:
@msfroh, @reta  I was thinking another way to do this, is via index setting rather than a mapping parameter. what you guys think? The setting will be like enable recovery_source or not. Either the recovery_source will be there for all fields or it will be there for no fields.


I know this is contradicting my above comment. The difference is earlier it was a mapping option and now it will be a proper index settings

Reply:
> I know this is contradicting my above comment. The difference is earlier it was a mapping option and now it will be a proper index settings

Hm ... I think the mapping should be the place to stay cohesive with `_source` property (which is managed through mappings).

Reply:
Thanks @reta

Reply:
Probably out of scope for this PR, but if we have a situation where we recover partial documents (because fields were removed from `_source` and `_recovery_source`), it might be nice to add a searchable meta field to the resulting doc (like `"_partial_recovery":true`).

That way, users could easily track down any partial docs and redrive them from another data source.

Reply:
@msfroh this is really good idea. I will work on this next, and I have some more ideas to ensure that partial recovery is not happening ever.